src/middleware/index.ts

```diff
@@ -39,11 +42,55 @@ export function expressRateLimiter(
      */
     // TODO: Throw ValidationError if schema is invalid
     const typeWeightObject = buildTypeWeightsFromSchema(schema, typeWeightConfig);
+
     // TODO: Throw error if connection is unsuccessful
-    const redisClient = new Redis(redisClientOptions); // Default port is 6379 automatically
+    const redisClient = connect(redisClientOptions); // Default port is 6379 automatically
     const rateLimiter = setupRateLimiter(rateLimiterAlgo, rateLimiterOptions, redisClient);
 
-    // return the rate limiting middleware
+    // stores request IDs to be processed
+    const requestQueue: { [index: string]: string[] } = {};
+
+    // Manages processing of event queue
+    const requestEvents = new EventEmitter();
+
+    /**
+     * Throttle rateLimiter.processRequest based on user IP to prevent inaccurate redis reads
+     * Throttling is based on a event driven promise fulfillment approach.
+     * Each time a request is received a promise is added to the user's request queue. The promise "subscribes"
+     * to the previous request in the user's queue then calls processRequest and resolves once the previous request
+     * is complete.
+     * @param userId
+     * @param timestamp
+     * @param tokens
+     * @returns
+     */
+    async function throttledProcess(
+        userId: string,
+        timestamp: number,
+        tokens: number
+    ): Promise<RateLimiterResponse> {
+        // Alternatively use crypto.randomUUID() to generate a random uuid
+        const requestId = `${timestamp}${tokens}`;
+
+        if (!requestQueue[userId]) {
+            requestQueue[userId] = [];
+        }
+        requestQueue[userId].push(requestId);
+
+        return new Promise((resolve) => {
+            requestEvents.once(requestId, async () => {
+                // process the request
+                const response = await rateLimiter.processRequest(userId, timestamp, tokens);
+                requestQueue[userId] = requestQueue[userId].slice(1);
+                // trigger the next event
+                requestEvents.emit(requestQueue[userId][0]);
+
+                resolve(response);
+            });
+            requestEvents.emit(requestQueue[userId][0]);
```

Comment:
The only critic I have, based on my understanding, is that if a request is received and there is nothing in queue (most cases) then we are setting an event listener and immediately triggering that event. This would put the request in the micro-task queue while other code runs before reaching out to Redis (which would put the request in the microtask queue again). This is a lot of potential waiting when there technically isn't anything to wait for, because there is no queue for this ip address.

You could solve this by adding a check that if the queue only has this one request, skip the event listening and go straight to Redis. When finished, remove the request for the queue and emit the event for the next one.